### PR TITLE
do not compute ordered watched resources on every push

### PR DESCRIPTION
**Pull request by ramaraochavali**
OrderedWatchedResources and KnownTypeUrls is computed on every push even though they change very rarely. This PR stores it in the proxy.

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Dual Stack
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [X] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Review by hzxuzhonghu — COMMENTED** [review]
How much time will it cost for the limited typeUrl? Looking at `orderWatchedResources`, it can be improved to reduce one loop.

File: pilot/pkg/model/context.go [hunk 1]
```diff
@@ -369,6 +369,12 @@ type Proxy struct {
 	// WatchedResources contains the list of watched resources for the proxy, keyed by the DiscoveryRequest TypeUrl.
 	WatchedResources map[string]*WatchedResource
 
+	// OrderedWatchedResources contains the watched resources for the proxy, ordered in accordance with known push order.
+	OrderedWatchedResources []*WatchedResource
+
+	// KnownTypeUrls contains the set of known type URLs for the proxy.
+	KnownTypeUrls sets.String
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/context.go:376]: We donot want to store all the results that can be used.

Comment (ramaraochavali) [reply pilot/pkg/model/context.go:376]: What do you mean? This is just a set of type urls - which would be a maximum of 5/6 in this list

**Comment by ramaraochavali** [comment]
> How much time will it cost for the limited typeUrl? Looking at orderWatchedResources, it can be improved to reduce one loop.

It does not take significant time (should be negligible) but allocates array for every push which this PR tries to avoid.,

**Comment by hzxuzhonghu** [comment]
It allocates array with at most 5~6  length

**Comment by ramaraochavali** [comment]
> It allocates array with at most 5~6 length

Yes. But why should we allocate and compute every time if we can store

**Comment by hzxuzhonghu** [comment]
Store it looks redundant and bit weired here, because we have storred too many watched resources

**Review by hzxuzhonghu — COMMENTED** [review]
I have a sketch: and we donot need calling ordered i think.

first iterate the pushOrders and push the watched resources in order, then push other resources

**Comment by ramaraochavali** [comment]
It is just a pointer to the existing watched resources

**Comment by ramaraochavali** [comment]
> first iterate the pushOrders and push the watched resources in order, then push other resources

Order is precomputed so that we can avoid lock during pushXds.

**Comment by howardjohn** [comment]
> Yes. But why should we allocate and compute every time if we can store

Seems like a premature optimization. An xds push allocates 1000s of objects; we shouldn't add complexity to prevent 1 such allocation?

**Comment by ramaraochavali** [comment]
>  An xds push allocates 1000s of objects; we shouldn't add complexity to prevent 1 such allocation?

Fair point. This was one of my comments in the other PR https://github.com/istio/istio/pull/48586#pullrequestreview-1799920472

**Comment by ramaraochavali** [comment]
Closing this - let us see if BMs show any significant improvement.

**Pull request closed without merging**